File: packages/seoos/analysis/voice.py
"""Measured voice: what a body of text actually does, not what it feels like.

Everything here is computed from the text. No model judgement, no adjectives
invented to fill a report. That matters because the product's whole claim on
tone is that a recommendation to change it comes with evidence attached, and
an LLM asked "is this tone right" will always produce a confident answer
whether or not it has grounds for one.

The fingerprint is deliberately boring: counts, ratios and variances. The
interesting part is the comparison. A single site's numbers mean very little;
the same numbers next to the five pages that currently outrank it mean a lot.
"""

from __future__ import annotations

import re
import statistics
from dataclasses import dataclass, field

_SENTENCE = re.compile(r"[.!?]+(?:\s|$)")
_WORD = re.compile(r"[A-Za-z']+")

# First person plural is the "we are a company" voice. Second person is the
# "here is what you do" voice. The ratio between them is the single most
# visible difference between a brochure and a useful page.
_WE = {"we", "our", "ours", "us", "ourselves"}
_YOU = {"you", "your", "yours", "yourself", "yourselves"}
_I = {"i", "my", "mine", "me", "myself"}

# Hedges make a sentence survivable in a legal review and useless to a reader.
_HEDGES = {
    "may", "might", "could", "perhaps", "possibly", "generally", "typically",
    "usually", "often", "sometimes", "arguably", "relatively", "fairly",
    "somewhat", "seemingly", "apparently", "potentially", "largely", "tends",
    "suggests", "appears", "likely", "presumably", "virtually", "essentially",
}

# The vocabulary that signals a page was written to fill a slot rather than to
# tell someone something. Shared with the humanizer's banned list on purpose:
# one definition of the tell, used by the checker and by the writer.
_FILLER = {
    "leverage", "seamless", "robust", "unlock", "elevate", "delve",
    "synergy", "holistic", "bespoke", "cutting-edge", "best-in-class",
    "world-class", "game-changing", "revolutionary", "innovative",
    "empower", "streamline", "optimize", "optimise", "utilize", "utilise",
    "facilitate", "transformative", "paradigm", "ecosystem", "landscape",
    "journey", "solution", "solutions", "offering", "offerings",
}

# Words that carry a fact. A page dense in these is telling you something;
# a page empty of them is describing itself.
_CONCRETE = re.compile(
    r"\b(\d+(?:\.\d+)?%|\$\d|₹\d|\d{4}|\d+(?:\.\d+)?\s?(?:x|times|hours?|days?|"
    r"weeks?|months?|years?|minutes?|seconds?|users?|customers?|clients?))\b",
    re.I,
)

_PASSIVE = re.compile(
    r"\b(?:is|are|was|were|be|been|being|get|got)\s+(?:\w+ly\s+)?(\w+ed|"
    r"built|made|done|given|taken|seen|known|shown|held|sent|kept|left|"
    r"found|told|brought|written|driven|chosen)\b",
    re.I,
)

_DASH = re.compile(r"[—–]")


@dataclass
class VoiceFingerprint:
    """The measured shape of a body of text."""

    words: int = 0
    sentences: int = 0
    paragraphs: int = 0

    sentence_len_mean: float = 0.0
    sentence_len_stdev: float = 0.0
    sentence_len_max: int = 0
    # Uniform sentence length is the second most reliable machine-writing
    # tell after the dash. A human varies; a generator does not.
    rhythm: float = 0.0

    reading_grade: float = 0.0
    lexical_density: float = 0.0

    we_per_1k: float = 0.0
    you_per_1k: float = 0.0
    i_per_1k: float = 0.0
    address: str = "impersonal"

    hedges_per_1k: float = 0.0
    filler_per_1k: float = 0.0
    concrete_per_1k: float = 0.0
    passive_ratio: float = 0.0
    question_ratio: float = 0.0
    imperative_ratio: float = 0.0
    dash_count: int = 0

    measured: bool = True
    notes: list[str] = field(default_factory=list)

    def as_dict(self) -> dict:
        return {
            "words": self.words,
            "sentences": self.sentences,
            "paragraphs": self.paragraphs,
            "sentence_len_mean": round(self.sentence_len_mean, 1),
            "sentence_len_stdev": round(self.sentence_len_stdev, 1),
            "sentence_len_max": self.sentence_len_max,
            "rhythm": round(self.rhythm, 2),
            "reading_grade": round(self.reading_grade, 1),
            "lexical_density": round(self.lexical_density, 3),
            "we_per_1k": round(self.we_per_1k, 1),
            "you_per_1k": round(self.you_per_1k, 1),
            "i_per_1k": round(self.i_per_1k, 1),
            "address": self.address,
            "hedges_per_1k": round(self.hedges_per_1k, 1),
            "filler_per_1k": round(self.filler_per_1k, 1),
            "concrete_per_1k": round(self.concrete_per_1k, 1),
            "passive_ratio": round(self.passive_ratio, 3),
            "question_ratio": round(self.question_ratio, 3),
            "imperative_ratio": round(self.imperative_ratio, 3),
            "dash_count": self.dash_count,
            "measured": self.measured,
            "notes": self.notes,
        }


def _syllables(word: str) -> int:
    return max(len(re.findall(r"[aeiouy]{1,2}", word.lower())), 1)


def fingerprint(text: str) -> VoiceFingerprint:
    """Measure one body of text. Below 120 words the ratios are noise, so the
    fingerprint says so rather than reporting a number nobody should trust."""
    fp = VoiceFingerprint()
    text = (text or "").strip()
    if not text:
        fp.measured = False
        fp.notes.append("No text to measure.")
        return fp

    words = _WORD.findall(text)
    fp.words = len(words)
    if fp.words < 120:
        fp.measured = False
        fp.notes.append(
            f"Only {fp.words} words. Voice ratios need about 120 before they "
            "mean anything, so these are not reported as measured."
        )

    sentences = [s for s in _SENTENCE.split(text) if s.strip()]
    fp.sentences = len(sentences)
    fp.paragraphs = len([p for p in text.split("\n\n") if p.strip()])

    lengths = [len(_WORD.findall(s)) for s in sentences if _WORD.findall(s)]
    if lengths:
        fp.sentence_len_mean = statistics.fmean(lengths)
        fp.sentence_len_stdev = statistics.pstdev(lengths) if len(lengths) > 1 else 0.0
        fp.sentence_len_max = max(lengths)
        # Coefficient of variation. Human prose usually lands between 0.45 and
        # 0.8; generated prose clusters far tighter than that.
        fp.rhythm = fp.sentence_len_stdev / fp.sentence_len_mean if fp.sentence_len_mean else 0.0

    if fp.words and lengths:
        syll = sum(_syllables(w) for w in words)
        fp.reading_grade = max(
            0.0,
            0.39 * (fp.words / len(lengths)) + 11.8 * (syll / fp.words) - 15.59,
        )
        fp.lexical_density = len({w.lower() for w in words}) / fp.words

    def per_1k(n: int) -> float:
        return (n / fp.words) * 1000 if fp.words else 0.0

    lower = [w.lower() for w in words]
    fp.we_per_1k = per_1k(sum(1 for w in lower if w in _WE))
    fp.you_per_1k = per_1k(sum(1 for w in lower if w in _YOU))
    fp.i_per_1k = per_1k(sum(1 for w in lower if w in _I))
    fp.hedges_per_1k = per_1k(sum(1 for w in lower if w in _HEDGES))
    fp.filler_per_1k = per_1k(sum(1 for w in lower if w in _FILLER))
    fp.concrete_per_1k = per_1k(len(_CONCRETE.findall(text)))

    if fp.you_per_1k > fp.we_per_1k * 1.4:
        fp.address = "reader-facing"
    elif fp.we_per_1k > fp.you_per_1k * 1.4:
        fp.address = "company-facing"
    elif fp.we_per_1k + fp.you_per_1k < 3:
        fp.address = "impersonal"
    else:
        fp.address = "balanced"

    if sentences:
        fp.passive_ratio = len(_PASSIVE.findall(text)) / len(sentences)
        fp.question_ratio = text.count("?") / len(sentences)
        starts = [s.strip().split()[0].lower() for s in sentences if s.strip().split()]
        imperatives = sum(
            1 for s in starts
            if s in {"use", "add", "check", "read", "start", "stop", "pick",
                     "write", "run", "open", "send", "make", "take", "try",
                     "set", "keep", "put", "call", "book", "get", "build"}
        )
        fp.imperative_ratio = imperatives / len(starts) if starts else 0.0

    fp.dash_count = len(_DASH.findall(text))
    return fp


# What a difference has to be before it is worth telling a client about. Set
# from the spread you see across ordinary business writing: anything smaller
# is inside the noise and would make the tool sound certain about nothing.
_MATERIAL = {
    "reading_grade": 2.0,
    "sentence_len_mean": 4.0,
    "rhythm": 0.12,
    "we_per_1k": 4.0,
    "you_per_1k": 4.0,
    "hedges_per_1k": 4.0,
    "filler_per_1k": 3.0,
    "concrete_per_1k": 3.0,
    "passive_ratio": 0.12,
}

_MEANING = {
    "reading_grade": ("reads harder than", "reads easier than"),
    "sentence_len_mean": ("uses longer sentences than", "uses shorter sentences than"),
    "rhythm": ("varies sentence length more than", "is more uniform than"),
    "we_per_1k": ("talks about itself more than", "talks about itself less than"),
    "you_per_1k": ("addresses the reader more than", "addresses the reader less than"),
    "hedges_per_1k": ("hedges more than", "hedges less than"),
    "filler_per_1k": ("uses more marketing filler than", "uses less marketing filler than"),
    "concrete_per_1k": ("carries more specifics than", "carries fewer specifics than"),
    "passive_ratio": ("uses more passive voice than", "uses less passive voice than"),
}


def compare(site: VoiceFingerprint, rivals: list[VoiceFingerprint]) -> dict:
    """Compare one site's voice against the pages it competes with.

    Returns only differences big enough to act on, each carrying the two
    numbers behind it. A recommendation without both numbers is an opinion,
    and this product does not ship opinions as findings.
    """
    usable = [r for r in rivals if r.measured]
    if not site.measured:
        return {
            "measured": False,
            "reason": "Not enough text on the client's own pages to measure a voice.",
            "differences": [],
            "sample_size": len(usable),
        }
    if len(usable) < 3:
        return {
            "measured": False,
            "reason": (
                f"Only {len(usable)} comparable pages could be measured. Three is "
                "the floor for calling a difference real rather than one rival's habit."
            ),
            "differences": [],
            "sample_size": len(usable),
        }

    differences = []
    for metric, threshold in _MATERIAL.items():
        mine = getattr(site, metric)
        theirs = [getattr(r, metric) for r in usable]
        median = statistics.median(theirs)
        gap = mine - median
        if abs(gap) < threshold:
            continue
        higher, lower = _MEANING[metric]
        differences.append({
            "metric": metric,
            "site": round(mine, 3),
            "rival_median": round(median, 3),
            "rival_range": [round(min(theirs), 3), round(max(theirs), 3)],
            "gap": round(gap, 3),
            "reads_as": higher if gap > 0 else lower,
            "direction": "higher" if gap > 0 else "lower",
        })

    differences.sort(key=lambda d: abs(d["gap"] / _MATERIAL[d["metric"]]), reverse=True)
    return {
        "measured": True,
        "sample_size": len(usable),
        "differences": differences,
        "verdict": (
            "No material difference from the pages you compete with."
            if not differences else
            f"{len(differences)} measured differences from the pages you compete with."
        ),
    }

File: apps/web/src/engine/voice.ts
/**
 * Measured voice, in the browser.
 *
 * A port of `packages/seoos/analysis/voice.py`, deliberately kept line for
 * line where it can be, because the two engines must not disagree about what
 * a rhythm of 0.34 means. Everything here is arithmetic over text: no model
 * call, no key, no network. That matters because the product's whole claim on
 * tone is that a recommendation to change how a company writes arrives with
 * evidence attached, and a model asked "is this tone right" will always
 * produce a confident answer whether or not it has grounds for one.
 *
 * The two refusals below are worth more than any number the module produces:
 * under 120 words the ratios are noise, and under three readable rivals there
 * is no field to compare against.
 */

const SENTENCE_SPLIT = /[.!?]+(?:\s|$)/;
const WORD = /[A-Za-z']+/g;

/* First person plural is the "we are a company" voice. Second person is the
   "here is what you do" voice. The ratio between them is the most visible
   difference between a brochure and a page worth reading. */
const WE = new Set(["we", "our", "ours", "us", "ourselves"]);
const YOU = new Set(["you", "your", "yours", "yourself", "yourselves"]);
const I = new Set(["i", "my", "mine", "me", "myself"]);

/* Hedges make a sentence survivable in a legal review and useless to a reader. */
const HEDGES = new Set([
  "may", "might", "could", "perhaps", "possibly", "generally", "typically",
  "usually", "often", "sometimes", "arguably", "relatively", "fairly",
  "somewhat", "seemingly", "apparently", "potentially", "largely", "tends",
  "suggests", "appears", "likely", "presumably", "virtually", "essentially",
]);

/* The vocabulary that signals a page written to fill a slot. Shared with the
   humanizer's banned list on purpose: one definition of the tell, used by the
   checker and by the writer, so the two cannot drift apart. */
const FILLER = new Set([
  "leverage", "seamless", "robust", "unlock", "elevate", "delve",
  "synergy", "holistic", "bespoke", "cutting-edge", "best-in-class",
  "world-class", "game-changing", "revolutionary", "innovative",
  "empower", "streamline", "optimize", "optimise", "utilize", "utilise",
  "facilitate", "transformative", "paradigm", "ecosystem", "landscape",
  "journey", "solution", "solutions", "offering", "offerings",
]);

/* Words that carry a fact. A page dense in these is telling you something; a
   page empty of them is describing itself. */
const CONCRETE =
  /\b(\d+(?:\.\d+)?%|[$£€]\d|\d{4}|\d+(?:\.\d+)?\s?(?:x|times|hours?|days?|weeks?|months?|years?|minutes?|seconds?|users?|customers?|clients?))\b/gi;

const PASSIVE =
  /\b(?:is|are|was|were|be|been|being|get|got)\s+(?:\w+ly\s+)?(?:\w+ed|built|made|done|given|taken|seen|known|shown|held|sent|kept|left|found|told|brought|written|driven|chosen)\b/gi;

const DASH = /[—–]/g;

const IMPERATIVES = new Set([
  "use", "add", "check", "read", "start", "stop", "pick", "write", "run",
  "open", "send", "make", "take", "try", "set", "keep", "put", "call",
  "book", "get", "build",
]);

export type Address = "reader-facing" | "company-facing" | "impersonal" | "balanced";

export type VoiceFingerprint = {
  label: string;
  url: string | null;
  words: number;
  sentences: number;
  paragraphs: number;
  sentenceLenMean: number;
  sentenceLenStdev: number;
  sentenceLenMax: number;
  /** Coefficient of variation of sentence length. Human prose: roughly 0.45 to 0.8. */
  rhythm: number;
  readingGrade: number;
  lexicalDensity: number;
  wePer1k: number;
  youPer1k: number;
  iPer1k: number;
  address: Address;
  hedgesPer1k: number;
  fillerPer1k: number;
  concretePer1k: number;
  passiveRatio: number;
  questionRatio: number;
  imperativeRatio: number;
  dashCount: number;
  measured: boolean;
  notes: string[];
};

/** Below this, the ratios are arithmetic rather than evidence. */
export const WORD_FLOOR = 120;
/** Below this many readable rivals, there is no field to compare against. */
export const RIVAL_FLOOR = 3;

function syllables(word: string): number {
  const m = word.toLowerCase().match(/[aeiouy]{1,2}/g);
  return Math.max(m ? m.length : 0, 1);
}

function mean(xs: number[]): number {
  return xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : 0;
}

/** Population standard deviation, matching Python's statistics.pstdev. */
function pstdev(xs: number[]): number {
  if (xs.length < 2) return 0;
  const m = mean(xs);
  return Math.sqrt(mean(xs.map((x) => (x - m) ** 2)));
}

function median(xs: number[]): number {
  if (!xs.length) return 0;
  const s = [...xs].sort((a, b) => a - b);
  const mid = s.length >> 1;
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
}

function empty(label: string, url: string | null): VoiceFingerprint {
  return {
    label, url,
    words: 0, sentences: 0, paragraphs: 0,
    sentenceLenMean: 0, sentenceLenStdev: 0, sentenceLenMax: 0, rhythm: 0,
    readingGrade: 0, lexicalDensity: 0,
    wePer1k: 0, youPer1k: 0, iPer1k: 0, address: "impersonal",
    hedgesPer1k: 0, fillerPer1k: 0, concretePer1k: 0,
    passiveRatio: 0, questionRatio: 0, imperativeRatio: 0, dashCount: 0,
    measured: false, notes: [],
  };
}

/**
 * Measure one body of text. Below 120 words the fingerprint reports
 * `measured: false` and says why, rather than printing a number nobody
 * should trust.
 */
export function fingerprint(text: string, label = "", url: string | null = null): VoiceFingerprint {
  const fp = empty(label, url);
  const body = (text || "").trim();
  if (!body) {
    fp.notes.push("No text to measure.");
    return fp;
  }

  const words = body.match(WORD) ?? [];
  fp.words = words.length;
  fp.measured = fp.words >= WORD_FLOOR;
  if (!fp.measured) {
    fp.notes.push(
      `Only ${fp.words} words. Voice ratios need about ${WORD_FLOOR} before they mean anything, so these are not reported as measured.`,
    );
  }

  const sents = body.split(SENTENCE_SPLIT).filter((s) => s.trim());
  fp.sentences = sents.length;
  fp.paragraphs = body.split(/\n\s*\n/).filter((p) => p.trim()).length;

  const lengths = sents
    .map((s) => (s.match(WORD) ?? []).length)
    .filter((n) => n > 0);

  if (lengths.length) {
    fp.sentenceLenMean = mean(lengths);
    fp.sentenceLenStdev = pstdev(lengths);
    fp.sentenceLenMax = Math.max(...lengths);
    fp.rhythm = fp.sentenceLenMean ? fp.sentenceLenStdev / fp.sentenceLenMean : 0;

    const syll = words.reduce((n, w) => n + syllables(w), 0);
    fp.readingGrade = Math.max(
      0,
      0.39 * (fp.words / lengths.length) + 11.8 * (syll / fp.words) - 15.59,
    );
    fp.lexicalDensity = new Set(words.map((w) => w.toLowerCase())).size / fp.words;
  }

  const per1k = (n: number) => (fp.words ? (n / fp.words) * 1000 : 0);
  const lower = words.map((w) => w.toLowerCase());
  const count = (set: Set<string>) => lower.filter((w) => set.has(w)).length;

  fp.wePer1k = per1k(count(WE));
  fp.youPer1k = per1k(count(YOU));
  fp.iPer1k = per1k(count(I));
  fp.hedgesPer1k = per1k(count(HEDGES));
  fp.fillerPer1k = per1k(count(FILLER));
  fp.concretePer1k = per1k((body.match(CONCRETE) ?? []).length);

  if (fp.youPer1k > fp.wePer1k * 1.4) fp.address = "reader-facing";
  else if (fp.wePer1k > fp.youPer1k * 1.4) fp.address = "company-facing";
  else if (fp.wePer1k + fp.youPer1k < 3) fp.address = "impersonal";
  else fp.address = "balanced";

  if (sents.length) {
    fp.passiveRatio = (body.match(PASSIVE) ?? []).length / sents.length;
    fp.questionRatio = (body.match(/\?/g) ?? []).length / sents.length;
    const starts = sents
      .map((s) => s.trim().split(/\s+/)[0]?.toLowerCase() ?? "")
      .filter(Boolean);
    const imp = starts.filter((s) => IMPERATIVES.has(s)).length;
    fp.imperativeRatio = starts.length ? imp / starts.length : 0;
  }

  fp.dashCount = (body.match(DASH) ?? []).length;
  return fp;
}

/**
 * What a difference has to be before it is worth telling a client about, set
 * from the spread across ordinary business writing. Anything smaller is inside
 * the noise, and a tool that lists every tiny gap sounds certain about nothing.
 */
export const MATERIAL: Record<string, number> = {
  readingGrade: 2.0,
  sentenceLenMean: 4.0,
  rhythm: 0.12,
  wePer1k: 4.0,
  youPer1k: 4.0,
  hedgesPer1k: 4.0,
  fillerPer1k: 3.0,
  concretePer1k: 3.0,
  passiveRatio: 0.12,
};

const MEANING: Record<string, [string, string]> = {
  readingGrade: ["reads harder than", "reads easier than"],
  sentenceLenMean: ["uses longer sentences than", "uses shorter sentences than"],
  rhythm: ["varies sentence length more than", "is more uniform than"],
  wePer1k: ["talks about itself more than", "talks about itself less than"],
  youPer1k: ["addresses the reader more than", "addresses the reader less than"],
  hedgesPer1k: ["hedges more than", "hedges less than"],
  fillerPer1k: ["uses more marketing filler than", "uses less marketing filler than"],
  concretePer1k: ["carries more specifics than", "carries fewer specifics than"],
  passiveRatio: ["uses more passive voice than", "uses less passive voice than"],
};

const LABEL: Record<string, string> = {
  readingGrade: "Reading grade",
  sentenceLenMean: "Mean sentence length",
  rhythm: "Rhythm",
  wePer1k: "We, per 1,000 words",
  youPer1k: "You, per 1,000 words",
  hedgesPer1k: "Hedges, per 1,000 words",
  fillerPer1k: "Marketing filler, per 1,000 words",
  concretePer1k: "Specifics, per 1,000 words",
  passiveRatio: "Passive voice, per sentence",
};

export type Difference = {
  metric: string;
  label: string;
  site: number;
  rivalMedian: number;
  rivalRange: [number, number];
  gap: number;
  readsAs: string;
  direction: "higher" | "lower";
};

export type VoiceComparison = {
  measured: boolean;
  reason: string | null;
  sampleSize: number;
  differences: Difference[];
  verdict: string | null;
};

/**
 * Compare one site's voice against the pages it competes with. Returns only
 * differences big enough to act on, each carrying both numbers behind it. A
 * recommendation without both numbers is an opinion, and this does not ship
 * opinions as findings.
 */
export function compare(site: VoiceFingerprint, rivals: VoiceFingerprint[]): VoiceComparison {
  const usable = rivals.filter((r) => r.measured);

  if (!site.measured) {
    return {
      measured: false,
      reason: "Not enough text on your own pages to measure a voice. About 120 words on a page is the floor.",
      sampleSize: usable.length,
      differences: [],
      verdict: null,
    };
  }
  if (usable.length < RIVAL_FLOOR) {
    return {
      measured: false,
      reason:
        `Only ${usable.length} comparable page${usable.length === 1 ? "" : "s"} could be measured. ` +
        `${RIVAL_FLOOR} is the floor for calling a difference real rather than one rival's habit.`,
      sampleSize: usable.length,
      differences: [],
      verdict: null,
    };
  }

  const differences: Difference[] = [];
  for (const metric of Object.keys(MATERIAL)) {
    const threshold = MATERIAL[metric];
    const mine = (site as unknown as Record<string, number>)[metric];
    const theirs = usable.map((r) => (r as unknown as Record<string, number>)[metric]);
    const med = median(theirs);
    const gap = mine - med;
    if (Math.abs(gap) < threshold) continue;
    const [higher, lower] = MEANING[metric];
    differences.push({
      metric,
      label: LABEL[metric],
      site: mine,
      rivalMedian: med,
      rivalRange: [Math.min(...theirs), Math.max(...theirs)],
      gap,
      readsAs: gap > 0 ? higher : lower,
      direction: gap > 0 ? "higher" : "lower",
    });
  }

  differences.sort((a, b) => Math.abs(b.gap / MATERIAL[b.metric]) - Math.abs(a.gap / MATERIAL[a.metric]));

  return {
    measured: true,
    reason: null,
    sampleSize: usable.length,
    differences,
    verdict: differences.length
      ? `${differences.length} measured difference${differences.length === 1 ? "" : "s"} from the pages you compete with.`
      : "No material difference from the pages you compete with.",
  };
}
